include/sparrow/mp_utils.hpp

```diff
@@ -424,8 +426,44 @@ namespace sparrow::mpl
     template <class T>
     concept boolean_like = std::is_assignable_v<std::add_lvalue_reference_t<std::decay_t<T>>, bool> and 
                            requires { static_cast<bool>(std::declval<T>()); };
-
+    
     /// Matches range types From whose elements are convertible to elements of range type To.
     template <class From, class To>
     concept convertible_ranges = std::convertible_to<std::ranges::range_value_t<From>, std::ranges::range_value_t<To>>;
+                   
+    // Matches any `unique_ptr` instance.
+    template <typename T>
+    concept unique_ptr = mpl::is_type_instance_of_v<std::remove_reference_t<T>, std::unique_ptr>;
+
+    // Matches any `unique_ptr` or derived instance.
+    template <typename T>
+    concept unique_ptr_or_derived = unique_ptr<T>
+                                    || std::derived_from<T, std::unique_ptr<typename T::element_type>>;
+
+    // Matches any `shared_ptr` instance.
+    template <typename T>
+    concept shared_ptr = mpl::is_type_instance_of_v<std::remove_reference_t<T>, std::shared_ptr>;
+
+    // Matches any `shared_ptr` or derived instance.
+    template <typename T>
+    concept shared_ptr_or_derived = shared_ptr<T>
+                                    || std::derived_from<T, std::shared_ptr<typename T::element_type>>;
+
+    // Matches any `unique_ptr` or `shared_ptr` instance.
+    template <typename T>
+    concept smart_ptr = unique_ptr<T> || shared_ptr<T>;
+
+    // Matches any `unique_ptr` or `shared_ptr` or derived instance.
+    template <typename T>
+    concept smart_ptr_and_derived = shared_ptr_or_derived<T> || unique_ptr_or_derived<T>;
+
+    static_assert(std::is_same_v<int, int> == true);
+
+    // Matches any type that has an element_type member.
+    template <typename T>
+    concept has_element_type = requires { typename T::element_type; };
+
+    // Matches any type that has a boolean operator.
+    template <class T>
+    concept has_boolean_operator = requires(T t) { bool(t); };
```

Comment:
Was `std::convertible_to<bool>` not sufficient?

Reply:
nope, because it requires implicit conversion.